File: OnlineJudge/LeetCode/codes/763.java
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Scanner;

class Solution {
  public List<Integer> partitionLabels(String S) {
    List<Integer> res = new ArrayList<>();
    List<Integer> u = new ArrayList<Integer>(Collections.nCopies(26, -1));

    int sz = S.length();
    for (int i = 0; i < sz; i++) {
      u.set(S.charAt(i) - 'a', i);
    }

    int l = 0, r = u.get(S.charAt(0) - 'a');
    for (int i = 0; i < sz; i++) {
      r = Math.max(r, u.get(S.charAt(i) - 'a'));
      if (r == i) {
        res.add(r - l + 1);
        l = i + 1;
      }
    }
    return res;
  }

  public static void main(String[] args) throws FileNotFoundException {
    FileInputStream fin = new FileInputStream("./assets/fipt.txt");
    PrintStream fot = new PrintStream(new FileOutputStream("./assets/fopt.txt"));

    System.setIn(fin);
    System.setOut(fot);

    Solution sol = new Solution();

    sol.input();
  }

  public void input() {
    Scanner in = new Scanner(System.in);
    while (in.hasNext()) {
      n = in.nextInt();

      for (int i = 0; i < n; i++) {
        a = in.next();
      }

      solve(a);
    }

    in.close();
  }

  public void solve(String x) {
    List<Integer> res;
    res = partitionLabels(x);
    otput(res);

  }

  public void otput(List<Integer> x) {
    System.out.println(x);
  }

  private int n;
  private String a = null;
}

File: OnlineJudge/LeetCode/codes/763.cpp
#include <algorithm>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <iostream>
#include <map>
#include <queue>
#include <stack>
#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<int> partitionLabels(string S) {
    vector<int> res;
    int u[26] = {0};

    int sz = S.size();
    for (int i = 0; i < sz; i++) {
      u[S[i] - 'a'] = i;
    }

    int l = 0, r = u[S[0] - 'a'];
    for (int i = 0; i < sz; i++) {
      r = max(r, u[S[i] - 'a']);
      if (r == i) {
        res.push_back(r - l + 1);
        l = i + 1;
      }
    }
    return res;
  }

  void input(void) {
    while (~scanf("%d", &n)) {
      for (int i = 0; i < n; i++) {
        cin >> a;
        solve(a);
      }
    }
  }

  void solve(string x) {
    vector<int> res;
    res = partitionLabels(x);
    otput(res);
  }
  void otput(vector<int> x) {
    for (int i = 0; i < x.size(); i++) cout << x[i] << endl;
  }

 private:
  int n;
  string a;
};

int main() {
  freopen("./assets/fipt.txt", "r", stdin);
  freopen("./assets/fopt.txt", "w", stdout);

  Solution sol;

  sol.input();

  return 0;
}
